biscuit-quote/src/lib.rs

```diff
@@ -456,83 +456,184 @@ impl Builder {
     }
 
     fn validate(&self) -> Result<(), error::LanguageError> {
-        if self.datalog_parameters == self.macro_parameters {
+        if self.macro_parameters.is_subset(&self.datalog_parameters) {
             Ok(())
         } else {
             let unused_parameters: Vec<String> = self
                 .macro_parameters
                 .difference(&self.datalog_parameters)
                 .cloned()
                 .collect();
-            let missing_parameters: Vec<String> = self
-                .datalog_parameters
-                .difference(&self.macro_parameters)
-                .cloned()
-                .collect();
             Err(error::LanguageError::Parameters {
-                missing_parameters,
+                missing_parameters: Vec::new(),
                 unused_parameters,
             })
         }
     }
 }
 
-impl ToTokens for Builder {
-    fn to_tokens(&self, tokens: &mut TokenStream) {
-        let (param_names, param_values): (Vec<String>, Vec<Expr>) =
-            self.parameters.clone().into_iter().unzip();
+struct Item {
+    parameters: HashSet<String>,
+    start: TokenStream,
+    middle: TokenStream,
+    end: TokenStream,
+}
 
-        let facts_quote = self.facts.iter().map(|f| {
-            quote! {
-                let mut fact = #f;
-                #(fact.set_macro_param(#param_names, #param_values).unwrap();)*
-                builder.add_fact(fact).unwrap();
-            }
-        });
+impl Item {
+    fn fact(fact: &Fact) -> Self {
+        Self {
+            parameters: fact
+                .parameters
+                .iter()
+                .flatten()
+                .map(|(name, _)| name.to_owned())
+                .collect(),
+            start: quote! {
+                let mut __biscuit_auth_item = #fact;
```

Comment:
These prefixed names are necessary, so that in

```rust
biscuit!("foo({item}, {builder})", item = 42, builder = "foo");
```

the builder or constructed item aren't shadowing user parameters.

Reply:
right, that makes sense